### DISCUSSION TITLE
First Class Polymorphism and Boolean Unification

### DISCUSSION THREAD

**Discussion by Superstar64** [flix/flix#6970] [Ideas]
As requested from the Gitter, How to combine first class polymorphism and boolean unification:

# Hindley Milner vs System-F
Hindley Milner is a type system where it is possible to do complete type inference without any type annotations, but it has limited expressability. Dually, System-F (first class polymorphism) is a much more expressive type system, but it's impossible to perform complete type inference without any type annotations.

There many systems that combine Hindley Milner and System-F. I would recommend the related work section of [FreezeML](https://homepages.inf.ed.ac.uk/slindley/papers/freezeml.pdf) and [Quick Look](https://www.microsoft.com/en-us/research/uploads/prod/2020/01/quick-look-icfp20-fixed.pdf) for a good overview of the current state of the research.

# Mixed Prefix Unification
A common necessity these systems have is [mixed prefix unification](https://www.lix.polytechnique.fr/Labo/Dale.Miller/papers/jsc92.pdf) (fair warning, this paper is somewhat dated and hard to read). Mixed prefix unification is a generalization of normal unification that allows you to have rigid type variables that don't unify with unification variable in a higher level. They need this to be able to unify higher rank types and first class polymorphism[^1]. Let me go over some mixed prefix unification examples then generalize them to boolean unification. I should note that I'm going subset of mixed unification that rejects everything that needs higher order unification.

## Example 1 (Normal Unification)
```
∀ a b. ∃ X Y. { a -> X = Y -> b }
```
This is just normal unification that you would expect to see in normal Hindley Milner. Here `X = b` and `Y = a`.

## Example 2 (Swaped Qualifiers)
```
∃ X Y. ∀ a b. { a -> X = Y -> b }
```
This is just like the previous example except the existential and universal qualifiers swapped. This example would error out (again without higher order unification) because `X` is trying to unify with `b`, which is a variable is a higher level. These sort of equations are the cause of skolemization errors that you commonly see in Haskell when using RankNTypes.

## Example 3 (Forall)
```
{ (forall a b. a -> b) = (forall c d. c -> d) }
∀ e f. { e -> f = e -> f}
```
This example shows why mixed prefix unification is needed in the first place. It's used when you want to unify forall types. You just simply pick fresh variables and substitute them into the qualify then you augment them as a fresh skolem variable.

## Example 4 (No Higher Order Unification)
```
∃ E. { (forall a. a) = (forall a. E) }
∃ E. ∀a. ( a = E }
```
This last example shows something that could be solved with higher order unification but that is typically rejected for simplicity.

## Constant Elimination
As I already mentioned, for mixed prefix unification to work, you need to make sure you don't allow a variable to unify with a constant at a higher level. This is normally something you just deal with when implementing it but can actually convert these problems into normal unification. This idea was inspired by a [conversation on the Theory Chatroom in the Programming Languages Discord](https://discord.com/channels/530598289813536771/530604668905193493/1083995977662345227) (and special thanks to @mb64 and @aradarbel10 for this idea). What you do is pick two types that you know are different and you duplicate your problem set substituting one type to one universal qualifier respectively.

Lets say here `Int` and `Char` are chosen as two types that we know are different. Here's how you would do this with the second example:
```
∃ X Y. ∀ a b. { a -> X, Y -> b }
∃ X Y. ∀ b. { Int -> X = Y -> b, Char -> X = Y -> b }
∃ X Y. { Int -> X = Y -> Int, Char -> X = Y -> Int, Int -> X = Y -> Char, Char -> X = Y -> Char }
```
As expected this would error out. Trying to unify a variable with a higher constant is translated to the variable being equal to two types at once.

Now for the translation of the third example:
```
{ (forall a b. a -> b) = (forall c d. c -> d) }
∀ e f. { e -> f = e -> f}
∀ f. { Int -> f = Int -> f, Char -> f = Char -> f}
{ Int -> Int = Int -> Int, Char -> Char = Char -> Char}
```
Here as expected this doesn't error out. The variable is never unified with a higher level constant, so gets translated into redundant equations.

# Mixed Prefix Boolean Unification
Luckily for boolean unification there are always two booleans that we know aren't equal (`true` and `false`). So know let me take all this together and use this to solve a unification problem with a higher rank type that involves boolean unification:
```
∀f. ∃ X. { (forall a. f a) = (forall a. f (a & X)) }
∀f. ∃ X. ∀a. { f a = f (a & X) }
∀f. ∃ X. { f False = f (False & X), f True = f (True & X) }
∀f. ∃ X. { False = False & X, True = True & X }
∀f. ∃ X. { False = False, True = X }
∀f. ∃ X. { X = True }
```

In summary, if Flix ever wants first class polymorphism. Your probably going to need an augmented form of mixed prefix unification that deletes constants for boolean equations via this duplication method I showed.

[^1]: Higher rank types are a subset of System-F (first class polymophism), where the `forall`s are not wrapped by type constructors. IE `(forall a. a) -> ...` but not `List (forall a. a)`.

**Comment by magnus-madsen** [comment]
@Superstar64 Thanks for the detailed write-up. It will probably take a bit for me to grok all the details.

Meanwhile, do we have any good use cases? I guess one use case would be for higher-ranked effect polymorhism? Or perhaps higher-ranked nullability? But I cannot yet imagine how such program examples would look like.

**Reply by Superstar64** [reply]
So, aside from all the usual uses of higher rank types, with regions specifically they allow you to implement region scoping in userspace (as a library function). Here's how this could look like:
```scala
// name taken from "Monadic Regions"
def letRGN(run[r2] : () -> a \ {r, r2}) : a \ {r} = ...

// taken from the flix documention
def sort(l: List[a]): List[a] with Order[a] =
    letRGN ( () -> {
        let arr = List.toArray(rc, l);
        Array.sort!(arr);
        Array.toList(arr)
    })
```

If you want an real example, in [Aith](github.com/superstar64/aith), I'm currently using this for borrowing a unique pointer as a shared mutable one. This is implemented completely in userland using an unsafe cast and zero cost via my staging system. Here's what it currently looks like:
```
inline borrow<
    A:region,
    B:region,
    C:region,
    M:boxed,
    J:*representation,
    K:multiplicity,
    N:pretype<J, K>
> = \(x : unique M in A)
 -> \(f : <D:region> M @ D in false -* N in C | B & D) 
 -> {
    let x = cast x : M @ false in A;
    let e = (f) @_ {x};
    (e, cast x : unique M in false)
};

increment(p :: unique int*) :: unique int* {
    let (x,p) = borrow {p} {
        <A : region> \(p :* int* @ A in false) -> {
            *p = (*p + 1)
        }
    };
    p
}
```

**Reply by JonathanStarup** [reply]
A simple one i've seen is carrying a function around that produces regioned data like `Int32 -> Region[r] -> Array[Int32, r]`. It is not very useful unless `r` is universally quantified

**Comment by Superstar64** [comment]
I feel I should preserve the conversation that inspired this method. Here is a manual transition (so some meta data is missing) of it in case the Discord link breaks. This conversation happened on 3/11/2023.

[mixed-prefix.txt](https://github.com/user-attachments/files/23431048/mixed-prefix.txt)
